### PR TITLE
feat: PKCE authentication flow for web logins #9890

### LINKED ISSUE #9890: Support PKCE for web logins
# Summary

The CLI currently supports PKCE. The web interface should do so as well.

# Proposal

I believe the API server will have to maintain a cache of code verifiers.

### PR DESCRIPTION
**Pull request by Marvin9**
Closes #9890 

This PR implements browser only PKCE flow for sso authentication.

It adds new attribute to the oidc config to enable the feature

```yaml
# argocd-cm
oidc.config: |
      name: dex
      issuer: http://localhost:4000/api/dex
      clientID: argo-cd-pkce
      requestedScopes: ["openid", "profile", "email", "groups"]
      enablePKCEAuthentication: true
```

### Flow in UI if PKCE enabled

Login:
- gets oidc configuration - clientID, issuer, scopes
- generates random code verifier
- calculate code challenge with code verifier
- save code verifier in browser session
- redirects to Idp with code challenge and other required metadata

consent screen callback to - `<argocd>/pkce/verify` - again thats the UI page which then get authorization code and asks Idp for user token
- extract `code` from query parameters
- get code verifier from browser session
- asks authorization server for token - with authorization code + code verifier
- set `id_token` in cookie

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines. - `not sure about this one`
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

**Comment by Marvin9** [comment]
@crenshaw-dev @jessesuen Can you please take a look or add relevant reviewers?

**Comment by jessesuen** [comment]
@Marvin9 I see that the example config adds the option to `oidc.config`:

```yaml
# argocd-cm
oidc.config: |
      name: dex
      issuer: http://localhost:4000/api/dex
      clientID: argo-cd-pkce
      requestedScopes: ["openid", "profile", "email", "groups"]
      enablePKCEAuthentication: true
```

Does this imply we don't support PKCE for someone using `dex.config`?

**Comment by Marvin9** [comment]
> @Marvin9 I see that the example config adds the option to `oidc.config`:
> 
> ```yaml
> # argocd-cm
> oidc.config: |
>       name: dex
>       issuer: http://localhost:4000/api/dex
>       clientID: argo-cd-pkce
>       requestedScopes: ["openid", "profile", "email", "groups"]
>       enablePKCEAuthentication: true
> ```
> 
> Does this imply we don't support PKCE for someone using `dex.config`?

I forgot that case. Please confirm that I understood this correct, what you are saying is the configuration of **only** OIDC from dex.config right?

```yaml
data:
  url: "https://argocd.example.com"
  dex.config: |
    connectors:
      # OIDC
      - type: oidc
        id: oidc
        name: OIDC
        config:
          issuer: https://example-OIDC-provider.com
          clientID: aaaabbbbccccddddeee
          clientSecret: $dex.oidc.clientSecret
```

**Comment by jessesuen** [comment]
What I meant was: when users use the bundled dex, today we generate a dex config for them like:

```yaml
staticClients:
- id: argo-cd
  name: Argo CD
  redirectURIs:
  - https://cd.demo.akuity.io/auth/callback
  secret: xxxxxxxxxxxxxxxxxxxxx        # this is automatically generated for them
- id: argo-cd-cli
  name: Argo CD CLI
  public: true
  redirectURIs:
  - http://localhost
  - http://localhost:8085/auth/callback
```

In order to support pkce web login when used with Argo CD's bundled dex, we need to:
1. set `public: true` for the `argo-cd` entry (just like we do for the `argo-cd-cli` entry)
2. not set a secret for the `argo-cd` entry

However, this behavior needs to be controlled with a new knob.  As discussed offline, I think for this PR, support for dex.config can be considered out of scope.

File: util/dex/config.go [hunk 1]
```diff
@@ -63,6 +63,14 @@ func GenerateDexConfigYAML(argocdSettings *settings.ArgoCDSettings, disableTls b
 			redirectURL,
 		},
 	}
+	argoCDPKCEStaticClient := map[string]interface{}{
```
Comment (Marvin9) [inline util/dex/config.go:66]: @jessesuen forgot to tell that I have added static client as well for PKCE

**Review by rpelczar — COMMENTED** [review]
UI LGTM

**Review by alexmt — APPROVED** [review]

**Pull request merged**